### PR TITLE
privilege/privileges: MySQLPrivilege caches mysql privilege table

### PR DESCRIPTION
**Pull request by tiancaiamao**
Add a MySQLPrivilege struct, it's the in-memory representation of MySQL user/db/tables_priv/columns_priv tables.

I'll use it to implement `privilege.Checker` interface.

### CODE REVIEW

File: privilege/privileges/cache_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	defer se.Close()
+	mustExec(c, se, "use mysql;")
+	mustExec(c, se, "truncate table tables_priv")
+
+	mustExec(c, se, `INSERT INTO mysql.tables_priv VALUES ("%", "db", "user", "table", "grantor", "2017-01-04 16:33:42.235831", "Grant,Index,Alter", "Insert,Update")`)
+
+	var p privileges.MySQLPrivilege
+	err = p.LoadTablesPrivTable(se)
+	c.Assert(err, IsNil)
+	c.Assert(p.TablesPriv[0].Host, Equals, "%")
```
Comment (ngaut) [inline privilege/privileges/cache_test.go]: Please use ` instead of "

File: privilege/privileges/cache_test.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	mustExec(c, se, "use mysql;")
+	mustExec(c, se, "truncate table columns_priv")
+
+	mustExec(c, se, `INSERT INTO mysql.columns_priv VALUES ("%", "db", "user", "table", "column", "2017-01-04 16:33:42.235831", "Insert,Update")`)
+	mustExec(c, se, `INSERT INTO mysql.columns_priv VALUES ("127.0.0.1", "db", "user", "table", "column", "2017-01-04 16:33:42.235831", "Select")`)
+
+	var p privileges.MySQLPrivilege
+	err = p.LoadColumnsPrivTable(se)
+	c.Assert(err, IsNil)
+	c.Assert(p.ColumnsPriv[0].Host, Equals, "%")
```
Comment (ngaut) [inline privilege/privileges/cache_test.go]: Ditto

**Comment by tiancaiamao** [comment]
PTAL @shenli

File: privilege/privileges/cache.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type userTableRow struct {
```
Comment (hanfei1991) [inline privilege/privileges/cache.go]: Is it more reasonable to call it `Cols` instead of `Row` ?

File: privilege/privileges/cache.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (hanfei1991) [inline privilege/privileges/cache.go]: This year is 2017

File: privilege/privileges/cache.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	DB         string
+	User       string
+	TableName  string
+	Grantor    string
+	Timestamp  time.Time
+	TablePriv  mysql.PrivilegeType
+	ColumnPriv mysql.PrivilegeType
+}
+
+const tablePrivMask = mysql.SelectPriv | mysql.InsertPriv | mysql.UpdatePriv | mysql.DeletePriv | mysql.CreatePriv | mysql.DropPriv | mysql.GrantPriv | mysql.IndexPriv | mysql.AlterPriv
```
Comment (hanfei1991) [inline privilege/privileges/cache.go]: You may put every `Mask` on top of the `TableRow` struct.

File: privilege/privileges/cache.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	return p.loadTable(ctx, "select * from mysql.tables_priv", p.decodeTablesPrivTableRow)
+}
+
+// LoadColumnsPrivTable loads the mysql.columns_priv table from database.
+func (p *MySQLPrivilege) LoadColumnsPrivTable(ctx context.Context) error {
+	return p.loadTable(ctx, "select * from mysql.columns_priv", p.decodeColumnsPrivTableRow)
+}
+
+func (p *MySQLPrivilege) loadTable(ctx context.Context, sql string,
+	decodeTableRow func(*ast.Row, []*ast.ResultField) error) error {
```
Comment (hanfei1991) [inline privilege/privileges/cache.go]: Maybe `tableRowDecoder` is better ?

File: privilege/privileges/cache.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+	p.ColumnsPriv = append(p.ColumnsPriv, value)
+	return nil
+}
+
+func decodeSetToPrivilege(s types.Set) (mysql.PrivilegeType, error) {
+	var ret mysql.PrivilegeType
+	for _, str := range strings.Split(s.Name, ",") {
+		priv, ok := mysql.SetStr2Priv[str]
+		if !ok {
+			return ret, errInvalidPrivilegeType.Gen("Unknown PrivilegeType!")
```
Comment (hanfei1991) [inline privilege/privileges/cache.go]: Why use Gen ?

File: privilege/privileges/cache.go (hunk 4 above)
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: it's 2016 when this PR was pushed.

File: privilege/privileges/cache.go (hunk 6 above)
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: By the name, `tableRowDecoder` means an object, while `decodeTableRow`  means an action
An action is better here, see line https://github.com/pingcap/tidb/pull/2388/files#diff-bf44fbe39866829e4b66d20ae4c5f501R141

File: privilege/privileges/cache.go (hunk 7 above)
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: It should not happen, table schema of `user` `db` `tables_priv` is controlled by us.
If the program goes to this line, something must be wrong...
In one word, this not a general purpose error information.

Comment (hanfei1991) [inline privilege/privileges/cache.go]: I mean it's enough to return errInvalidPrivilegeType.

File: privilege/privileges/cache.go (hunk 3 above)
Comment (tiancaiamao) [inline privilege/privileges/cache.go]: Maybe `Record` is a better name, I find it's exact what I want.

**Comment by tiancaiamao** [comment]
PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @shenli

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**